src/ClickTracker/HtmlClickTracker.php

```diff
@@ -54,9 +54,9 @@ public static function replaceHrefUrls($html, $replace) {
 
     // Find anything like href="..." or href='...' inside a tag.
     $tmp = preg_replace_callback(
-      ';(\<[^>]*href *= *")([^">]+)(");', $callback, $html);
+      ';(\<[^>]*href *= *")([^#">]+)(");', $callback, $html);
     return preg_replace_callback(
-      ';(\<[^>]*href *= *\')([^\'>]+)(\');', $callback, $tmp);
+      ';(\<[^>]*href *= *\')([^#\'>]+)(\');', $callback, $tmp);
```

Comment:
Am I reading correctly that this ignores links that have a `#` *anywhere* in the URL?

For example, suppose one composes a mailblast on behalf of `civicrm.org` which includes some discussion of a new feature and its documentation:

```html
<p>
CiviCRM v9 adds the "Relationship" tab!  Check out the
<a href="https://docs.civicrm.org/user/en/latest/organising-your-data/contacts/#relationships-tab">new docs</a>.
</p>
```

(This is by no means exclusive to `civicrm.org` docs - anchors are common enough on the web at large. It's just an example off the top of my head.)

If I wrote+sent a message like that, then it would be quite surprising (to me) for the link-tracking to be disabled just because my link included a `#`. After skimming some [docs](https://tools.ietf.org/html/rfc7231#section-7.1.2) and trying some throw-away PHP scripts, it seems entirely legal for a PHP script to send `Location: https://docs.civicrm.org/user/en/latest/organising-your-data/contacts/#relationships-tab`. So I don't see a technical reason to skip those URLs.

Aside: It might be good update `HtmlClickTrackerTest` to include some examples with anchors.

Reply:
you right @totten , the idea of this PR was to ignore from tracking only local anchors urls that start with `#` , no an absolute url with an anchor at the end..

I'll review the PR

Reply:
Any update on this Luciano?

Reply:
@sluc23 @MikeyMJCO - After re-reading this, I think it might be better to put the instruction somewhere else -- after the `$url` has been extracted. Adjusting this regex to do it properly feels scary.

If you trace what happens after the regex, these two spots look promising:

* Inside of the `$replace(string $url)` function ([code](https://github.com/civicrm/org.civicrm.flexmailer/blob/6a0e9782e7b3587c4fdd002fd8163e6b3cc6d86e/src/ClickTracker/HtmlClickTracker.php#L17-L25)). It has a similar skip rule for `{token}`-based URLs. 
* Inside of the `getTrackerURL(string $url,...)` function ([code](https://github.com/civicrm/civicrm-core/blob/cd999d6ba51900bae0986085c5b4478f91fb240b/CRM/Mailing/BAO/TrackableURL.php#L50-L56)). It has a similar skip rule for images+CSS URLs 

Skipping internal anchors might be as simple as `if ($url[0] === '#') { return $url; } `? (But I haven't tried it.)

The only difference between `$replace` and `getTrackerURL()`: the first is Flexmailer-only; the second is Flexmailer+MailingBAO.

I'm gonna close this PR but hope you get a chance to try out one of the other spots.